projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php

```diff
@@ -322,13 +322,36 @@ private function prepare_menu_item_icon( $icon ) {
 			if ( 0 === strpos( $icon, 'data:image/svg+xml' ) ) {
 				$img = $icon;
 			} elseif ( 0 === strpos( $icon, 'dashicons-' ) ) {
-				$img = sanitize_html_class( $icon );
+				$icon = $this->prepare_dashicon( $icon );
+				$img  = sanitize_html_class( $icon );
 			}
 		}
 
 		return $img;
 	}
 
+	/**
+	 * Prepares the dashicon for consumption by Calypso. If the dashicon isn't found in a list of known icons
+	 * we will return the default dashicon.
+	 *
+	 * @param string $icon The dashicon string to check.
+	 *
+	 * @return string If the dashicon exists in core we return the dashicon, otherwise we return the default dashicon.
+	 */
+	private function prepare_dashicon( $icon ) {
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$dashicon_list = include WP_CONTENT_DIR . '/mu-plugins/masterbar/admin-menu/dashicon-set.php';
+		} else {
+			$dashicon_list = include JETPACK__PLUGIN_DIR . '/modules/masterbar/admin-menu/dashicon-set.php';
+		}
+
+		if ( isset( $dashicon_list[ $icon ] ) ) {
```

Comment:
[Non-blocking] Should we also check the value of the array item? Seems the boolean value is otherwise useless and can cause confusion (someone can assume it indicates whether the dashicon is enabled/disabled).
```suggestion
		if ( isset( $dashicon_list[ $icon ] ) && $dashicon_list[ $icon ] ) {
```

Reply:
> Should we also check the value of the array item?

I wouldn't check the value. The boolean value is indeed useless. Defining a set as:
`
array(
   'setItem1' => true,
   'setItem2' => true,
   .
   .
   .
   'setItemN' => true,
);
`
Is the standardish way of defining a set when you don't have the [Set data structure](https://www.php.net/manual/en/class.ds-set.php) available.

> Seems the boolean value is otherwise useless and can cause confusion (someone can assume it indicates whether the dashicon is enabled/disabled).

If you find this confusing I could change it to an array of strings and perform an array search over the items. We would lose performance since array_search is O(n) time complexity and checking if an item exists in a set is O(1). But it shouldn't be a problem since the set only contains a couple of hundred elements.

Reply:
I think adding an additional condition for checking for true would be enough, even though it's redundant, it would future proof it against assumptions (e.g. someone unfamiliar with the changes might assume that they'll need to mark the keys as false to have them hidden).

Reply:
Done.